### PR TITLE
Create BaseTrackSetFeature parent for CrateFeature & BasePlaylistFeature

### PR DESCRIPTION
**Pull request by xeruf**
Create common BaseTrackCollectionFeature ancestor for crates & playlists - I will then move the common methods into that. Opening this PR already to see if you agree with the overall approach.

### CODE REVIEW

**Comment by uklotzde** [comment]
This PR deserves a better title, even if it is only a draft. The title of a PR should outline the essence of its changes, a summary distilled from all commit messages ;)

**Comment by Xerus2000** [comment]
@uklotzde you are right, I was tired and just wanted to get it out there before going to sleep ^^

File: src/library/playlistfeature.h [hunk 1]
```diff
@@ -1,6 +1,3 @@
-// playlistfeature.h
```
Comment (uklotzde) [inline src/library/playlistfeature.h:1]: If your remove those comments please also replace the legacy include guards with `#pragma once`

**Review by uklotzde — COMMENTED** [review]
Yes, this is how we need to extract common functionality that is duplicated.

It is comparable to what I started in #2538 to support new use cases while avoiding code duplication.

**Comment by uklotzde** [comment]
We need to think about the term *collection* since this is already used for the whole collection of all tracks. Both crates and playlists contain only a subset (playlist: duplicates allowed) or *selection* of such a collection.

**Comment by Holzhaus** [comment]
> We need to think about the term _collection_ since this is already used for the whole collection of all tracks. Both crates and playlists contain only a subset (playlist: duplicates allowed) or _selection_ out of such a collection.

*Selection* is also ambiguous though, because it could also refer to a visual selection in the tableview. I think *collection* fits best, we just need to use `Library` for the collection of all tracks instead.

**Comment by uklotzde** [comment]
@Holzhaus Exactly my thoughts when mentioning the term "selection" in this context. I just wasn't able to find a better distinction from "collection". Unfortunately in Mixxx the term "library" is used for both the whole database and also the top-level UI components.

We could rename this base class later after we agreed on a consistent terminology for all this library-related stuff.

**Comment by Xerus2000** [comment]
yikes, some of the code duplication is in the CrateFeatureHelper, let's see how to approach that...

File: src/library/playlistfeature.h (hunk 1 above)
Comment (Xerus2000) [reply src/library/playlistfeature.h:1]: this is the sort of thing that should just be done for all files at once with a commit to master tbh ^^^

Comment (Be-ing) [reply src/library/playlistfeature.h:1]: Mass code style changes cause unnecessary merge conflicts with open pull requests. We update style as we work instead.

Comment (Xerus2000) [reply src/library/playlistfeature.h:1]: In general I agree, but how would replacing these lines create a conflict?

Comment (uklotzde) [reply src/library/playlistfeature.h:1]: Please let's stop discussing minor details and get things done. We have to agree on certain processes.

File: CMakeLists.txt [hunk 2]
```diff
@@ -348,6 +348,7 @@ add_library(mixxx-lib STATIC EXCLUDE_FROM_ALL
   src/library/baseplaylistfeature.cpp
   src/library/basesqltablemodel.cpp
   src/library/basetrackcache.cpp
+  src/library/basetrackcollectionfeature.cpp
```
Comment (Be-ing) [inline CMakeLists.txt:351]: Please update build/depends.py until we remove scons

**Comment by Be-ing** [comment]
I don't like how easy it is to confuse BaseTrackCollectionFeature with TrackCollection either. How about BaseTrackSetFeature?

File: CMakeLists.txt (hunk 2 above)
Comment (Xerus2000) [reply CMakeLists.txt:351]: ah, wasn't aware of that - the cmake update was done automatically by clion - will do

**Comment by Xerus2000** [comment]
I searched through the whole codebase and found no reference to the `showPage` signal in the baseplaylistfeature, and it doesn't seem to be a Qt-thing either, so I guess I can remove it?

**Comment by Xerus2000** [comment]
I have just rebased the branch on master and now suddenly, no commit on the branch builds anymore. With that info, it has to be related to the first commit, and it seems to be, but it is super weird. I can't see any relevant change from master affecting it, and the first commit is a very simple refactor. I am not even sure how I can split it up any more, since it is already such a small and self-contained change.

On run, Mixxx exits immediately after opening with exit code 11:

image

**Comment by Xerus2000** [comment]
The problem might be that I use "this" in the constructor of PlaylistFeature and SetLogFeature to construct the PlaylistTableModel - but it worked before, so that is very weird...

**Comment by Xerus2000** [comment]
Okay, it works with safeMode, oof...

**Comment by Xerus2000** [comment]
```shell
/home/janek/daten/projects/_forks/mixxx/cmake-build-debug/mixxx --developer --settingsPath /var/tmp/mixxx-dev --safeMode
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "aif"
Info [Main]: SoundSourceProxy -   3 : "libsndfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "aiff"
Info [Main]: SoundSourceProxy -   2 : "libsndfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "caf"
Info [Main]: SoundSourceProxy -   1 : "libsndfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "flac"
Info [Main]: SoundSourceProxy -   3 : "Xiph.org libFLAC"
Info [Main]: SoundSourceProxy -   1 : "libsndfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "it"
Info [Main]: SoundSourceProxy -   2 : "MODPlug"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "med"
Info [Main]: SoundSourceProxy -   2 : "MODPlug"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "mod"
Info [Main]: SoundSourceProxy -   2 : "MODPlug"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "mp3"
Info [Main]: SoundSourceProxy -   2 : "MAD: MPEG Audio Decoder"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "ogg"
Info [Main]: SoundSourceProxy -   3 : "Xiph.org OggVorbis"
Info [Main]: SoundSourceProxy -   1 : "libsndfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "okt"
Info [Main]: SoundSourceProxy -   2 : "MODPlug"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "opus"
Info [Main]: SoundSourceProxy -   3 : "Xiph.org libopusfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "s3m"
Info [Main]: SoundSourceProxy -   2 : "MODPlug"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "stm"
Info [Main]: SoundSourceProxy -   2 : "MODPlug"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "wav"
Info [Main]: SoundSourceProxy -   2 : "libsndfile"
Info [Main]: SoundSourceProxy - SoundSource providers for file extension "xm"
Info [Main]: SoundSourceProxy -   2 : "MODPlug"
Debug [Main]: "Mixxx" "2.3.0-alpha-pre" "(git playlist-crate-refactor r7248; built on: Apr 10 2020 @ 14:10:10)" is starting...
Debug [Main]: Compile time library versions:
Debug [Main]: Qt: 5.14.1
Debug [Main]: libshout: 2.4.1
Debug [Main]: PortAudio: 1246720 PortAudio V19.6.0-devel, revision 396fe4b6699ae929d3a685b3ef8a7e97396139a4
Debug [Main]: RubberBand: 1.8.1
Debug [Main]: SoundTouch: 2.1.1
Debug [Main]: TagLib: 1.11.1
Debug [Main]: ChromaPrint: 1.4.3
Debug [Main]: Vorbis: Xiph.Org libVorbis 1.3.5
Debug [Main]: libsndfile: libsndfile-1.0.28
Debug [Main]: FLAC: 1.3.2
Debug [Main]: libmp3lame: 3.100
Debug [Main]: QStandardPaths::writableLocation(HomeLocation): "/home/janek"
Debug [Main]: QStandardPaths::writableLocation(DataLocation): "/home/janek/.local/share/Mixxx"
Debug [Main]: QCoreApplication::applicationDirPath() "/home/janek/daten/projects/_forks/mixxx/cmake-build-debug"
Debug [Main]: Loading resources from  "/home/janek/daten/projects/_forks/mixxx/cmake-build-debug/res/"
Debug [StatsManager]: StatsManager thread starting up.
Debug [Main]: Configuration file is at the current version 2.3.0-alpha-pre
Info [Main]: BroadcastSettings - Found 1 profile(s)
Debug [Main]: Loading resources from  "/home/janek/daten/projects/_forks/mixxx/cmake-build-debug/res/"
Debug [Main]: Found and will use default keyboard preset "/home/janek/daten/projects/_forks/mixxx/cmake-build-debug/res/keyboard/en_US.kbd.cfg"
Debug [Main]: Loading resources from  "/home/janek/daten/projects/_forks/mixxx/cmake-build-debug/res/"
Debug [Main]: Loading translations for locale "en_US" from translations folder "/home/janek/daten/projects/_forks/mixxx/cmake-build-debug/res/translations/" : success
Debug [Main]: Loading resources from  "/home/janek/daten/projects/_forks/mixxx/cmake-build-debug/res/"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Light" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Bold" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "SemiBold Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Bold Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "ExtraBold" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Regular" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "ExtraBold Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "SemiBold" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Light Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Light" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Bold" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "SemiBold Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Bold Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "ExtraBold" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Regular" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "ExtraBold Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "SemiBold" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Open Sans" style: "Light Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Light Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Bold" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Bold Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Medium" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Light" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Medium Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Standard" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Regular" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Light Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Bold" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Bold Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Medium" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Light" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Medium Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Standard" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Regular" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Debug [Main]: FONT LOADED family: "Ubuntu" style: "Italic" point sizes: "6,7,8,9,10,11,12,14,16,18,20,22,24,26,28,36,48,72"
Info [Main]: DbConnection - Available drivers for database connections: ("QSQLITE")
Debug [Main]: DbConnection - Opening database connection "MIXXX-1" QSqlDatabase(driver="QSQLITE", database="file:///var/tmp/mixxx-dev/mixxxdb.sqlite", host="", port=-1, user="mixxx", open=false)
Info [Main]: DbConnectionPool - Cloned thread-local database connection "MIXXX-1" QSqlDatabase(driver="QSQLITE", database="file:///var/tmp/mixxx-dev/mixxxdb.sqlite", host="", port=-1, user="mixxx", open=true)
Info [Main]: MixxxMainWindow - Connecting to database
Debug [Main]: DbConnectionPooled - Found thread-local database connection "MIXXX-1" QSqlDatabase(driver="QSQLITE", database="file:///var/tmp/mixxx-dev/mixxxdb.sqlite", host="", port=-1, user="mixxx", open=true)
Info [Main]: MixxxMainWindow - Initializing or upgrading database schema
Info [Main]: SchemaManager - Database schema is up-to-date at version 32
Debug [Main]: Compressor attack per frame:  0.000408163 decay per frame:  4.08163e-05
Debug [Main]: JACK client name set
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
Debug [Main]: EngineNetworkStream - addWorker: worker added
Debug [Main]: EngineNetworkStream - worker slots used: "2 out of 16"
Debug [Main]: BroadcastManager - addConnection: created connection for profile "Connection 1"
Warning [Main]: ControlDoublePrivate::getControl returning NULL for ( "[AutoDJ]" , "enabled" )
Debug [Main]: WARNING: AudioInput already registered!
Debug [Main]: WARNING: AudioInput already registered!
Debug [Main]: WARNING: AudioInput already registered!
Debug [Main]: WARNING: AudioInput already registered!
Info [Main]: GlobalTrackCache - Creating instance
Debug [Main]: DbConnectionPooled - Found thread-local database connection "MIXXX-1" QSqlDatabase(driver="QSQLITE", database="file:///var/tmp/mixxx-dev/mixxxdb.sqlite", host="", port=-1, user="mixxx", open=true)
Info [Main]: TrackCollection - Connecting database
Info [Main]: TrackCollectionManager - Starting library scanner thread
Debug [LibraryScanner 1]: LibraryScanner - Entering thread
Debug [LibraryScanner 1]: DbConnection - Opening database connection "MIXXX-2" QSqlDatabase(driver="QSQLITE", database="file:///var/tmp/mixxx-dev/mixxxdb.sqlite", host="", port=-1, user="mixxx", open=false)
Info [LibraryScanner 1]: DbConnectionPool - Cloned thread-local database connection "MIXXX-2" QSqlDatabase(driver="QSQLITE", database="file:///var/tmp/mixxx-dev/mixxxdb.sqlite", host="", port=-1, user="mixxx", open=true)
Debug [LibraryScanner 1]: DbConnectionPooled - Found thread-local database connection "MIXXX-2" QSqlDatabase(driver="QSQLITE", database="file:///var/tmp/mixxx-dev/mixxxdb.sqlite", host="", port=-1, user="mixxx", open=true)
Info [LibraryScanner 1]: LibraryScanner - Cleaning up database...
Info [LibraryScanner 1]: LibraryScanner - Finished database cleanup: 0 ms
Debug [LibraryScanner 1]: LibraryScanner - Event loop starting
Info [Main]: TrackCollection - Connecting track source
Debug [Main]: BaseTrackTableModel - Skipping header properties for unsupported column 21 "Channels"
Debug [Main]: BaseTrackTableModel - Skipping header properties for unsupported column 19 "Samplerate"
Debug [Main]: BaseTrackTableModel - Skipping header properties for unsupported column 40 "#"
Debug [Main]: BaseTrackTableModel - Skipping header properties for unsupported column 45 "Timestamp"
Debug [Main]: Committing transaction on "MIXXX-1" result: true
Debug [Main]: BaseTrackTableModel - Skipping header properties for unsupported column 21 "Channels"
Debug [Main]: BaseTrackTableModel - Skipping header properties for unsupported column 19 "Samplerate"

Process finished with exit code 11
```

It seems to break when switching branches, maybe it is some unrelated instability in the mixxx database?
Although, I have just created a completely new settingsPath, ran it in safeMode and developerMode and it still failed as above.

@Be-ing any ideas?

**Comment by Be-ing** [comment]
Have you tried running it in a debugger?

**Comment by Holzhaus** [comment]
Compile Mixxx in debug mode, then run:

```
gdb. /mixxx
```
then type `run`.

Reproduce the crash, then type `bt`.

**Comment by xerus2000** [comment]
> Have you tried running it in a debugger and getting a backtrace?

I didn't know how to debug C++ tbh - thanks @Holzhaus 
```
...
[New Thread 0x7ffedd7fa700 (LWP 1045)]
[New Thread 0x7ffedcff9700 (LWP 1046)]
Warning [Main]: ControlDoublePrivate::getControl returning NULL for ( "[Channel1]" , "intro_start_position" ) 
Warning [Main]: ControlDoublePrivate::getControl returning NULL for ( "[Channel1]" , "intro_end_position" ) 

Thread 1 "mixxx" received signal SIGSEGV, Segmentation fault.
0x0000555555e9c50f in WaveformMarkRange::~WaveformMarkRange (this=0x555565cef590, __in_chrg=<optimized out>) at src/waveform/renderers/waveformmarkrange.cpp:18
warning: Source file is more recent than executable.
18          if (!m_activeColor.isValid()) {
```
I first thought this was caused by improper cleaning when switching branches, but even after completely rebuilding everything this error appears...

**Comment by Be-ing** [comment]
> warning: Source file is more recent than executable.

C++ debuggers rely on the source files being the same as the compiled executable you are debugging. Otherwise they will point you to the wrong line and you will get confused.

**Comment by xerus2000** [comment]
> C++ debuggers rely on the source files being the same as the compiled executable you are debugging. Otherwise they will point you to the wrong line and you will get confused.

Well, I am confused now anyways, what am I supposed to do? ^^ 
As said, even a full rebuild didn't seem to help.

**Comment by Holzhaus** [comment]
The warning indicates that your source files are *newer* than the compiled executable. Either you edited the source files without compiling them afterwards, or you're trying to debug the wrong the wrong executable (e.g. the system-installed `mixxx` instead of the one from the build dir).

**Comment by xerus2000** [comment]
Yes, I was able to interpret that, but didn't quite now what to do since I thought I had already done a full rebuild. Apparently the executable was from an older build and the current one was in `cmake-build-debug/mixxx` (default for IntelliJ)

This is the error when running it:
```
Thread 1 "mixxx" received signal SIGSEGV, Segmentation fault.
QObject::thread (this=this@entry=0x55555893d950) at kernel/qobject.cpp:1512
1512    kernel/qobject.cpp: No such file or directory.
```

**Comment by Holzhaus** [comment]
> Apparently the executable was from an older build and the current one was in `cmake-build-debug/mixxx` (default for IntelliJ)

IDEs try to reduce complexity, but more often than not they actually make stuff even more complex. I suggest to always compile on the command line.

The error you posted is not a complete backtrace. Compile Mixxx, run it in `gdb` and type `bt` when it crashes to get a list of function calls that lead to the crash. That will mKe it much easier to debug.

**Comment by Be-ing** [comment]
> Apparently the executable was from an older build and the current one was in cmake-build-debug/mixxx (default for IntelliJ)

I had to configure KDevelop to set the make target to "install" for the project to make sure the binary I built in KDevelop is the same one I run in a separate shell. Maybe you can do something similar for your IDE.

**Comment by xerus2000** [comment]
image

**Comment by Be-ing** [comment]
In my experience when I have gotten crashes in Qt code, it is because I passed an invalid value from application code, usually an invalid index to a container class. In this case I am guessing you tried to use an invalid QModelIndex. So start debugging from the last frame in application code (frame 4 in BaseTrackTableModel) and work from there.

**Comment by xerus2000** [comment]
> The problem might be that I use "this" in the constructor of PlaylistFeature and SetLogFeature to construct the PlaylistTableModel - but it worked before, so that is very weird...

Then that is very likely to be the issue. I'm still confused why it worked previously, but anyways, let's see how to change that properly...

**Comment by xerus2000** [comment]
@Be-ing how about merging this now before it becomes too big? I would then create a follow-up PR to actually factor out the individual methods.

File: CMakeLists.txt [hunk 3]
```diff
@@ -353,7 +353,7 @@ add_library(mixxx-lib STATIC EXCLUDE_FROM_ALL
   src/library/baseplaylistfeature.cpp
   src/library/basesqltablemodel.cpp
   src/library/basetrackcache.cpp
-  src/library/basetrackcollectionfeature.cpp
+  src/library/basetracksetfeature.cpp
```
Comment (Be-ing) [inline CMakeLists.txt:356]: update depends.py for scons

Comment (xerus2000) [reply CMakeLists.txt:356]: I'm sorry, fixed it :sweat_smile:

**Comment by xerus2000** [comment]
@Be-ing can we merge? :)

File: src/library/basetracksetfeature.cpp [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#include "basetracksetfeature.h"
+
+BaseTrackSetFeature::BaseTrackSetFeature(
+        Library* pLibrary,
+        UserSettingsPointer pConfig,
+        const QString& rootViewName)
+        : LibraryFeature(pLibrary, pConfig),
+        m_rootViewName(rootViewName){
```
Comment (uklotzde) [inline src/library/basetracksetfeature.cpp:8]: Please reformat all new code according to .clang-format

File: src/library/basetracksetfeature.h [hunk 5]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+#include "libraryfeature.h"
```
Comment (uklotzde) [inline src/library/basetracksetfeature.h:3]: Includes in the same directory should still be relative to the root source folder, i.e. "library/libraryfeature.h"

File: src/library/playlistfeature.h (hunk 1 above)
Comment (uklotzde) [reply src/library/playlistfeature.h:1]: Still waiting for the #pragma once change. Please don't resolve conversations if you don't follow the request.

**Review by uklotzde — CHANGES_REQUESTED** [review]
Some minor formal issues

File: src/library/playlistfeature.h (hunk 1 above)
Comment (xerus2000) [reply src/library/playlistfeature.h:1]: Sorry, I had done it on all files except the one you commented on :see_no_evil:

**Review by xeruf — COMMENTED** [review]
Looks like I missed a few details, fixed now :)

**Comment by Be-ing** [comment]
Sorry, I forgot to merge 2.2 to master after merging #2617. There were some merge conflicts I had to fix and now this branch has merge conflicts with master. Please fix those.

**Comment by xerus2000** [comment]
done :)

**Comment by Holzhaus** [comment]
I'll review and test this tomorrow

File: src/library/missingtablemodel.h [hunk 6]
```diff
@@ -1,13 +1,13 @@
 #ifndef MISSINGTABLEMODEL_H
 #define MISSINGTABLEMODEL_H
```
Comment (Holzhaus) [inline src/library/missingtablemodel.h:2]: Please replace this with `#pragma once`.

**Review by Holzhaus — CHANGES_REQUESTED** [review]
Thanks. This works fine for me. I left some comments below.

File: src/library/setlogfeature.h [hunk 7]
```diff
@@ -3,26 +3,26 @@
...
-#include <QSqlTableModel>
 #include <QAction>
+#include <QLinkedList>
 #include <QPointer>
+#include <QSqlTableModel>
 
 #include "library/baseplaylistfeature.h"
 #include "preferences/usersettings.h"
 
 class SetlogFeature : public BasePlaylistFeature {
```
[the remark this answers is not in the document]
Comment (xerus2000) [reply src/library/setlogfeature.h:14]: I'm not sure I am quite ready to do that yet, I understand the class in general but I feel like I'm missing some quirks

File: src/library/baseplaylistfeature.cpp [hunk 8]
```diff
@@ -1,152 +1,140 @@
...
         const QString& rootViewName)
-        : LibraryFeature(pLibrary, pConfig),
+        : BaseTrackSetFeature(pLibrary, pConfig, rootViewName),
           m_playlistDao(pLibrary->trackCollections()->internalCollection()->getPlaylistDAO()),
-          m_pPlaylistTableModel(nullptr),
-          m_rootViewName(rootViewName) {
+          m_pPlaylistTableModel(pModel) {
+    pModel->setParent(this);
+
+    initActions();
```
[the remark this answers is not in the document]
Comment (xerus2000) [reply src/library/baseplaylistfeature.cpp:33]: for consistency with the cratefeature

[the remark this answers is not in the document]
Comment (xerus2000) [reply src/library/baseplaylistfeature.cpp:33]: this is basically preparation for the following pull request, where some of these actions will move into basetracksetfeature

**Comment by xerus2000** [comment]
Can we finally merge? :sweat_smile:  As said I will create a follow-up PR immediately, so we can resolve minor issues later. But I need this as foundation.

File: src/library/basetracksetfeature.cpp [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+        Library* pLibrary,
+        UserSettingsPointer pConfig,
+        const QString& rootViewName)
+        : LibraryFeature(pLibrary, pConfig),
+          m_rootViewName(rootViewName) {
+}
+
+void BaseTrackSetFeature::activate() {
+    emit switchToView(m_rootViewName);
+    emit disableSearch();
```
Comment (Be-ing) [inline src/library/basetracksetfeature.cpp:13]: We're disabling searching for crates??

Comment (Be-ing) [reply src/library/basetracksetfeature.cpp:13]: Oh I see this is how it was before...

Comment (xerus2000) [reply src/library/basetracksetfeature.cpp:13]: yep, this PR is a pure refactor, no functional changes (except for this minimal menu order change which was made for consistency: https://github.com/mixxxdj/mixxx/pull/2621/commits/2e835ddc783e34c1fd0dbbfbd23aecb1d3f000bb)

Comment (xerus2000) [reply src/library/basetracksetfeature.cpp:13]: that's why I want to get it merged, further refactoring is not unlikely to introduce functional changes

**Comment by uklotzde** [comment]
@xerus2000 Thanks for improving the code quality before adding new features.

**Pull request merged**